File: pathfinder-api/src/main/java/de/cubbossa/pathfinder/storage/cache/StorageCache.java
package de.cubbossa.pathfinder.storage.cache;

import com.google.common.base.Preconditions;

import java.util.Collection;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;

public interface StorageCache<E> {

  void write(E e);

  void invalidate(E e);

  void invalidateAll();

  record CacheCollection<K, V>(Collection<V> present, Collection<K> absent) {

    public static <K, V> CacheCollection<K, V> empty(Collection<K> absent) {
      Preconditions.checkNotNull(absent);
      return new CacheCollection<>(new HashSet<>(), absent);
    }
  }

  record CacheMap<K, V>(Map<K, V> present, Collection<K> absent) {

    public static <K, V> CacheMap<K, V> empty(Collection<K> absent) {
      Preconditions.checkNotNull(absent);
      return new CacheMap<>(new HashMap<>(), absent);
    }
  }
}

File: pathfinder-api/src/main/java/de/cubbossa/pathfinder/storage/cache/StorageCache.kt
package de.cubbossa.pathfinder.storage.cache

import com.google.common.base.Preconditions

interface StorageCache<E> {
    fun write(e: E)

    fun invalidate(e: E)

    fun invalidateAll()

    @JvmRecord
    data class CacheCollection<K, V>(val present: Collection<V>, val absent: Collection<K>) {
        companion object {
            fun <K, V> empty(absent: Collection<K>): CacheCollection<K, V> {
                Preconditions.checkNotNull(absent)
                return CacheCollection(HashSet(), absent)
            }
        }
    }

    @JvmRecord
    data class CacheMap<K, V>(val present: Map<K, V>, val absent: Collection<K>) {

        companion object {
            fun <K, V> empty(absent: Collection<K>): CacheMap<K, V> {
                Preconditions.checkNotNull(absent)
                return CacheMap(HashMap(), absent)
            }
        }
    }
}
